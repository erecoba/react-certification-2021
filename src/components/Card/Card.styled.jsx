import styled from 'styled-components';

const Container = styled.button`
  border-radius: 20px;
  border: 0;
  outline: 0;
  padding: 0;
  margin: 8px 5px;

  background-color: transparent;

  cursor: pointer;

  width: 345px;
  height: 245px;

  position: relative;
`;

const ImageFit = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
  border-radius: 20px;
`;

const GradientContainer = styled.div`
  width: 100%;
  height: 100px;
  position: absolute;
  left: 0;
  bottom: 0;
  border-radius: 0 0 20px 20px;
  background: linear-gradient(
    180deg,
    rgba(252, 254, 255, 0) 15%,
    rgba(41, 46, 50, 0.8) 70%
  );

  padding: 10px 16px;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
`;

const Title = styled.h2`
  font-size: 0.875rem;
  text-align: left;
  color: #fff;
`;

const Description = styled.span`
  font-size: 0.75rem;
  text-align: left;
  color: #fff;
  width: 100%;

  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const TagChannel = styled.div`
  position: absolute;
  right: 10px;
  top: 10px;
  border-radius: 30px;

  padding: 5px 15px;
  display: flex;
  justify-content: center;
  align-items: center;

  background-color: ${({ theme }) => theme.base};
  color: #fff;
`;

export { Container, ImageFit, GradientContainer, Title, Description, TagChannel };
