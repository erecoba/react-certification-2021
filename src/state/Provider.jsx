import React, { createContext, useContext, useReducer } from 'react';
import youtubeReducer, { initialState as youtubeInitialState } from './reducers/youtube';

const YoutubeContext = createContext({ ...youtubeInitialState });

const useYoutubeVideo = () => {
  const context = useContext(YoutubeContext);
  if (!context) {
    throw new Error("Can't use useYoutubeVideo with out a Provider");
  }
  return context;
};

const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(youtubeReducer, youtubeInitialState);

  return (
    <YoutubeContext.Provider value={{ state, dispatch }}>
      {children}
    </YoutubeContext.Provider>
  );
};

export { useYoutubeVideo };

export default Provider;
