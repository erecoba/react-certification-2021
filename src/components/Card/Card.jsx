import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';

import { FavoriteProcess } from '../../utils/favoriteProcess';
import StarIcon from '../../ui/icons/Star';
import { useSession } from '../../state/Provider';

import {
  Container,
  ImageFit,
  GradientContainer,
  Title,
  Description,
  TagChannel,
  FavStar,
} from './Card.styled';

const Card = ({
  id,
  title,
  description,
  imgUrl,
  channelTitle,
  isFluid,
  canFavorite,
  linkBase = 'videos',
}) => {
  const history = useHistory();
  const [isFavButtonShown, setIsFavButtonShown] = useState(false);
  const { sessionState } = useSession();

  const handleAddFavorite = () => {
    if (!sessionState.user) {
      // eslint-disable-next-line no-alert
      alert('Log in before to add a favorite video');
      return;
    }

    const videoFavorite = new FavoriteProcess({
      id,
      title,
      description,
      imgUrl,
      channelTitle,
    });

    videoFavorite.addOrRemoveVideo().updateLocalStorage().displayMessage();
  };

  return (
    <Container
      isFluid={isFluid}
      onMouseEnter={() => setIsShown(true)}
      onMouseLeave={() => setIsShown(false)}
      aria-label="card-wrapper"
    >
      <ImageFit
        src={imgUrl}
        alt={title}
        onClick={() => history.push(`/${linkBase}/${id}`)}
      />
      <GradientContainer
        aria-label="card-container"
        onClick={() => history.push(`/${linkBase}/${id}`)}
      >
        <Title>{title}</Title>
        <Description>{description}</Description>
      </GradientContainer>
      <TagChannel>{channelTitle}</TagChannel>
      {canFavorite && isShown && (
        <FavStar onClick={handleAddFavorite} aria-label="favorite-button">
          <StarIcon color="white" />
        </FavStar>
      )}
    </Container>
  );
};

export default Card;
