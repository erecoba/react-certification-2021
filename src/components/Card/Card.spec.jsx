import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';

import Provider from '../../state/Provider';

import Card from './Card';

jest.mock('../../state/actions/youtube');

describe('Avatar', () => {
  let props;
  let theme;
  window.alert = jest.fn();

  beforeAll(() => {
    theme = {
      colors: {
        white: '#fff',
      },
    };
    props = {
      title: '',
      description: '',
      imgUrl: '',
      channelTitle: '',
      id: 'sG9rX6Ifzhw',
      isFluid: false,
    };
  });

  it('should render correctly inside button', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <Card {...props} />
      </ThemeProvider>
    );

    expect(
      asFragment(
        <ThemeProvider theme={theme}>
          <Card {...props} />
        </ThemeProvider>
      )
    ).toMatchSnapshot();
  });

  it('should render the title', () => {
    props = { title: 'El título' };
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Card {...props} />
      </ThemeProvider>
    );

    expect(getByText('El título')).toBeTruthy();
  });

  it('should can be click to push a new url', () => {
    const wrapper = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Card {...props} />
        </ThemeProvider>
      </BrowserRouter>
    );

    const button = wrapper.getByLabelText('card-container');
    fireEvent.click(button);
  });

  it('should has the property width 100% while isFluid is true', () => {
    props.isFluid = true;
    const wrapper = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Card {...props} />
        </ThemeProvider>
      </BrowserRouter>
    );

    const button = wrapper.getByLabelText('card-container');
    expect(button).toHaveStyle('width: 100%');
  });

  it('should add a new video to favorite', () => {
    props.canFavorite = true;
    const wrapper = render(
      <BrowserRouter>
        <Provider>
          <ThemeProvider theme={theme}>
            <Card {...props} />
          </ThemeProvider>
        </Provider>
      </BrowserRouter>
    );

    const cardWrapper = wrapper.getByLabelText('card-wrapper');
    fireEvent.mouseEnter(cardWrapper);

    const favButton = wrapper.getByLabelText('favorite-button');
    fireEvent.click(favButton);
  });
});
