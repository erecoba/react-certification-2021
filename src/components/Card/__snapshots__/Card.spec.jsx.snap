// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render correctly inside button 1`] = `
<DocumentFragment>
  .c0 {
  border-radius: 20px;
  border: 0;
  outline: 0;
  padding: 0;
  margin: 8px 5px;
  background-color: transparent;
  cursor: pointer;
  width: 345px;
  height: 245px;
  position: relative;
}

.c1 {
  width: 100%;
  height: 100%;
  object-fit: cover;
  border-radius: 20px;
}

.c2 {
  width: 100%;
  height: 100px;
  position: absolute;
  left: 0;
  bottom: 0;
  border-radius: 0 0 20px 20px;
  background: linear-gradient( 180deg, rgba(252,254,255,0) 15%, rgba(41,46,50,0.8) 70% );
  padding: 10px 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  font-size: 0.875rem;
  text-align: left;
  color: #fff;
}

.c4 {
  font-size: 0.75rem;
  text-align: left;
  color: #fff;
  width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c5 {
  position: absolute;
  right: 10px;
  top: 10px;
  border-radius: 30px;
  padding: 5px 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #fff;
}

<button
    aria-label="card-container"
    class="c0"
  >
    <img
      alt=""
      class="c1"
      src=""
    />
    <div
      class="c2"
    >
      <h2
        class="c3"
      />
      <span
        class="c4"
      />
    </div>
    <div
      class="c5"
    />
  </button>
</DocumentFragment>
`;
